File: ControlLibraryFS/StringFormatFloatTextBlock.fs
namespace ControlLibraryFS

open System
open Avalonia
open Avalonia.Controls

type StringFormatFloatTextBlock() =
    inherit TextBlock()
    static let StringFormatProperty = AvaloniaProperty.Register<StringFormatFloatTextBlock, string>("StringFormat", "")
    static let ValueProperty        = AvaloniaProperty.Register<StringFormatFloatTextBlock, double>("Value")

    static let updateValue (e : AvaloniaPropertyChangedEventArgs<double>) =
        match e.Sender with
        | :? StringFormatFloatTextBlock as sender ->
            sender.Text <- String.Format(sender.StringFormat, e.NewValue.Value)
        | _ -> ()

    static let updateStringFormat (e : AvaloniaPropertyChangedEventArgs<string>) =
        match e.Sender with
        | :? StringFormatFloatTextBlock as sender ->
            sender.Text <- String.Format(e.NewValue.Value, sender.Value)
        | _ -> ()

    static do
        StringFormatFloatTextBlock.RegisterRenderDependenicies()
        StringFormatProperty.Changed |> Observable.subscribe updateStringFormat |> ignore
        ValueProperty.Changed        |> Observable.subscribe updateValue        |> ignore

    static member RegisterRenderDependenicies() =
        Visual.AffectsRender<StringFormatFloatTextBlock>(
            StringFormatProperty,
            ValueProperty
        )

    member this.StringFormat
        with get () = this.GetValue(StringFormatProperty)
        and  set v  = this.SetValue(StringFormatProperty, v) |> ignore

    member this.Value
        with get () : double  = this.GetValue(ValueProperty)
        and  set (v : double) = this.SetValue(ValueProperty, v) |> ignore

File: ControlLibraryCS/StringFormatFloatTextBlock.cs
namespace ControlLibraryCS
{
    using System;
    using Avalonia;
    using Avalonia.Controls;

    public class StringFormatFloatTextBlock : TextBlock
    {
        public StringFormatFloatTextBlock()
        {
        }

        public static readonly StyledProperty<string> StringFormatProperty = AvaloniaProperty.Register<StringFormatFloatTextBlock, string>("StringFormat", "");
        public static readonly StyledProperty<double> ValueProperty = AvaloniaProperty.Register<StringFormatFloatTextBlock, double>("Value");

        private static void UpdateValue(AvaloniaPropertyChangedEventArgs<double> e)
        {
            if (e.Sender is StringFormatFloatTextBlock sender)
            {
                sender.Text = string.Format(sender.StringFormat, e.NewValue.Value);
            }
        }

        private static void UpdateStringFormat(AvaloniaPropertyChangedEventArgs<string> e)
        {
            if (e.Sender is StringFormatFloatTextBlock sender)
            {
                sender.Text = string.Format(e.NewValue.Value, sender.Value);
            }
        }

        static StringFormatFloatTextBlock()
        {
            RegisterRenderDependenicies();
            StringFormatProperty.Changed.Subscribe(UpdateStringFormat);
            ValueProperty.Changed.Subscribe(UpdateValue);
        }

        private static void RegisterRenderDependenicies()
        {
            AffectsRender<StringFormatFloatTextBlock>(
                StringFormatProperty,
                ValueProperty
            );
        }

        public string StringFormat
        {
            get { return GetValue(StringFormatProperty); }
            set { SetValue(StringFormatProperty, value); }
        }

        public double Value
        {
            get { return GetValue(ValueProperty); }
            set { SetValue(ValueProperty, value); }
        }
    }
}